runtime/dynstats.c

```diff
@@ -499,12 +643,161 @@ dynstats_createCtr(dynstats_bucket_t *b, const uchar* metric, dynstats_ctr_t **c
 	RETiRet;
 }
 
+/* not supported in older platforms - disable */
+#ifndef O_CLOEXEC
+#define O_CLOEXEC 0
+#endif
+/*
+ *	Helper function adapted from imfile of the same name - possible candidate for refactor.
+ */
+static rsRetVal ATTR_NONNULL()
+writeStateFile(dynstats_bucket_t *b, const char *file_name, const char *content) {
+	const size_t toWrite = strlen(content);
+	ssize_t w = 0;
+	DEFiRet;
+	const int fd = open(file_name, O_CLOEXEC | O_NOCTTY | O_WRONLY | O_CREAT | O_TRUNC, 0600);
+	if(fd < 0) {
+		LogError(errno, RS_RET_IO_ERROR, "dynstats: cannot open state file '%s' for "
+			"persisting file state - some data will probably be duplicated "
+			"on next startup", file_name);
+		ABORT_FINALIZE(RS_RET_IO_ERROR);
+	}
+
+	w = write(fd, content, toWrite);
+	if(w != (ssize_t) toWrite) {
+		LogError(errno, RS_RET_IO_ERROR, "dynstats: partial write to state file '%s' "
+			"this may cause trouble in the future. We will try to delete the "
+			"state file, as this provides most consistent state", file_name);
+		unlink(file_name);
+		ABORT_FINALIZE(RS_RET_IO_ERROR);
+	}
+
```

Comment:
Yes, the initial idea for this feature, was simply to back up the counts, in the way imfile backs up the file positions. This way, at least if rsyslog is restarted normally, the counts are not reset. 

Making this feature completely durable, seems to add lots of overhead, which we could add, but does it make sense to make it configurable? If so, i'd like to push that as a future enhancement?

Reply:
Not fully durable impl is ok.
Configurable durability is ok too.
Since we are doing this in a different thread, I thought the overhead of `sync` wouldn't really be a problem.
I think its ok either way, but we should definitely document the behavior inline and in user facing documentation of dynstats.